### PR TITLE
Remove context transformers. Add provider hooks

### PR DESCRIPTION
**Pull request by justinabrahms**
refs https://github.com/open-feature/research/pull/22

### CODE REVIEW

File: specification/sections/02-providers.md
Comment (beeme1mr) [inline specification/sections/02-providers.md:116]:
```suggestion
Feature flag management systems often need to transform the context structures the user provides or generally address lifecycle concerns. A `provider hook` exposes a mechanism for providers to register [`Hooks`](./04-hooks.md) to tap into various stages of the flag evaluation lifecycle.
```

File: specification/sections/02-providers.md
Comment (beeme1mr) [inline specification/sections/02-providers.md:120]: I'm not sure if this needs to be a method or function. For example, in TypeScript it could just be a property on the class.

File: specification/sections/02-providers.md
Comment (beeme1mr) [inline specification/sections/02-providers.md:126]: private readonly hook: Hook[] = [new MyProviderHook()];

**Review by beeme1mr — APPROVED** [review]
LGTM

**Comment by beeme1mr** [comment]
Hey @justinabrahms, could you please signoff your commit?

**Comment by justinabrahms** [comment]
Signed off again.

File: specification/sections/02-providers.md
Comment (toddbaert) [inline specification/sections/02-providers.md:116]: I think context transformation is a key use case for provider hooks, so it's worth mentioning, however they are just as flexible as other hooks. I think this example should come after the description... So basically reverse the order of these 2 sentences.

File: specification/sections/02-providers.md
Comment (toddbaert) [inline specification/sections/02-providers.md:120]: I think we need to state that the provider abstraction MUST support attaching hooks, though obviously the provider author doesn't need to attach them to their provider. I think having some implementations not supporting provider hooks will be problematic, just too inconsistent.

**Review by toddbaert — COMMENTED** [review]
I think we need to modify 4.4.1 to include provider hooks... (Also, I'm not sure what it's trying to say when it mentions flag evaluation options)

File: specification/sections/02-providers.md
Comment (toddbaert) [inline specification/sections/02-providers.md:116]: I think context transformation is a key use case for provider hooks, so it's worth mentioning, however they are just as flexible as other hooks. If we're going to make this section about provider hooks generally,I think this example should come after the description... So basically reverse the order of these 2 sentences and highlight that context transformation is just one example use case.

**Comment by weyert** [comment]
So we are now adding support for provider hooks which appears to call `getProviderHooks` on the provider implementation. I am wondering what happens when someone would calls`OpenFeature.clearHooks()`? 

Would this also remove these provider hooks or only the user added hooks? 

Or is my assumption incorrect that the `getProviderHooks` will register the defined hooks in `OpenFeature`-singleton or are two separate things?

**Comment by justinabrahms** [comment]
@weyert Great call out. I would expect that this wouldn't remove provider hooks, which may be necessary for a provider to function.  Does that make sense to you?

**Comment by s-sen** [comment]
What is the use case of removing Provider hook? This hook is optionally supplied by the Provider and the app developer has no knowledge of it. Certain default features carried out by provider hook can be overridden by the app developer using hook hints if the provider chooses to document such option. But clear hook operation initiated by the app developer should not remove provider hooks.

**Comment by toddbaert** [comment]
It seems like we've basically identified application author-facing hooks and provider hooks. The clear hooks method is meant for the application author and probably shouldn't impact the provider hooks which the application author isn't aware of 

I'm not sure the clear hooks method is even in the spec (which IMO is fine).

**Review by toddbaert — APPROVED** [review]

**Review by skyerus — APPROVED** [review]

**Comment by beeme1mr** [comment]
Hey @justinabrahms, we should be good to merge this PR once all conversations have been resolved. Once this is merged, I plan on releasing v0.2.0.

**Review by thomaspoignant — APPROVED** [review]

**Comment by agentgonzo** [comment]
I'm late to the party here (I was off on holiday).

```
Hooks MUST be evaluated in the following order: - before: API, Client, Invocation, Provider - after: Provider, Invocation,
```
When talking about context transformation, why are we recommending that it is done via a hook? Why not just recommend having the provider layer do the transformation within the [Flag Evaluation stage](https://github.com/open-feature/spec/blob/main/specification/sections/01-flag-evaluation.md#flag-evaluation)?

As I understand it, the vender-specific format of the evaluation context is only needed by the vendor's SDK. As such, it (IMO) should be encapsulated within the provider layer. 

By transforming the context as a hook, we are introducing encapsulation leakage and passing the transformed context back to the OpenFeature SDK (only for it to be passed back into the provider layer by the [flag evaluation stage](https://github.com/open-feature/spec/blob/main/specification/sections/01-flag-evaluation.md#flag-evaluation)). 

Would it not be simpler (and better encapsulated) to leave the context transformation entirely within the flag evaluation stage of the provider?


WRT to the original messages on Slack about adding a provider hook to allow setting additional parameters to the context so that they can be logged by OpenTelemetry - that makes perfect sense and it's a good idea to add the vendor hooks as stated.

I'm just thinking that it's better to recommend transforming the context as part of the provider's Flag Evaluation stage rather than a hook.

**Comment by agentgonzo** [comment]
Additionally, I think it would be better to change the `MUST` for the provider hook to be a `MAY`. I don't like enforcing behaviour that may end up being a no-op.

If we change it to `MAY`, then we can say "if you provide a hook, it WILL be called", but if you don't provide a hook, then the OpenFeature SDK will just skip over to the next stage. This is easily provided by the SDK itself as a default no-op call.

@justinabrahms @beeme1mr ?

**Comment by toddbaert** [comment]
>By transforming the context as a hook, we are introducing encapsulation leakage and passing the transformed context back to the OpenFeature SDK (only for it to be passed back into the provider layer by the [flag evaluation stage](https://github.com/open-feature/spec/blob/main/specification/sections/01-flag-evaluation.md#flag-evaluation)).

This is a good observation, and I think I agree with your conclusion, and discussions yesterday with @beeme1mr and @justinabrahms kinda elucidated the same problem.

I think we can keep the ordering as is, since provider-related side effects should happen as close to the provider as possible, but we should remove the additional recommendation that context transformation be done in hooks.

> I'm just thinking that it's better to recommend transforming the context as part of the provider's Flag Evaluation stage rather than a hook.

Sounds like we're on the same page.

**Comment by justinabrahms** [comment]
> Additionally, I think it would be better to change the `MUST` for the provider hook to be a `MAY`. I don't like enforcing behaviour that may end up being a no-op.

In java, we're just providing a default noop implementation as part of the interface. I don't feel strongly on this wording.

**Comment by toddbaert** [comment]
https://github.com/open-feature/spec/pull/129 Here is the wording change, as well as a removal of the entire `context transformer` concept, which so far has only been adopted in the JS SDK.

**Pull request merged**
